（#9598 ）Compile Error

```
/home/ubuntu/azerothcore/modules/mod-autobalance/src/AutoBalance.cpp:60:21: fatal error: 'ABModuleScript' does not refer to a value
    FOR_SCRIPTS_RET(ABModuleScript, itr, end, ret) // return true by default if not scripts
/home/ubuntu/azerothcore/modules/mod-autobalance/src/AutoBalance.h:29:7: note: declared here
class ABModuleScript : public ModuleScript
1 error generated.
make[2]: *** [modules/CMakeFiles/modules.dir/build.make:300: modules/CMakeFiles/modules.dir/mod-autobalance/src/AutoBalance.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:1116: modules/CMakeFiles/modules.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

----------------------------------------------------------------------------------------------------------------------------------

```cmake
-- Ubuntu 20.04 LTS x86_64(Py3.7.9)

-- CMake version: 3.22.0
-- The CXX compiler identification is Clang 10.0.0
-- The C compiler identification is Clang 10.0.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Running cmake hook: AFTER_LOAD_CONF
-- No hooks registered for AFTER_LOAD_CONF
-- Enabled С++17 standard
-- Detected 64-bit platform
-- UNIX: Using jemalloc
-- UNIX: Using default configuration directory
-- UNIX: Using default library directory
-- UNIX: Configuring uninstall target
-- UNIX: Created uninstall target
-- UNIX: Detected compiler: /usr/bin/clang
-- Clang: Minimum version required is 10.0.0, found 10.0.0 - ok!
-- Performing Test CLANG_HAVE_PROPER_CHARCONV
-- Performing Test CLANG_HAVE_PROPER_CHARCONV - Success
-- Clang: All warnings enabled
-- Running cmake hook: AFTER_LOAD_CMAKE_MODULES
-- No hooks registered for AFTER_LOAD_CMAKE_MODULES
-- Using mysql-config: /usr/bin/mysql_config
-- Found MySQL library: /usr/lib/x86_64-linux-gnu/libmysqlclient.so
-- Found MySQL headers: /usr/include/mysql
-- Found MySQL executable: /usr/bin/mysql
-- Found git binary : /usr/bin/git

* AzerothCore revision            : fad214efe018+ 2021-12-10 15:05:28 -0300 (HEAD branch)
* AzerothCore buildtype           : RelWithDebInfo

* Install core to                 : /home/ubuntu/azeroth-server
* Install libraries to            : /home/ubuntu/azeroth-server/lib
* Install configs to              : /home/ubuntu/azeroth-server/etc

* Build world/auth                : Yes (default)
* Build with scripts              : Yes (static)
* Build with modules              : Yes (static)
* Build map/vmap tools            : No  (default)
* Build unit tests                : No  (default)
* Build core w/PCH                : Yes (default)
* Build scripts w/PCH             : Yes (default)
* Show all warnings               : Yes
* Use coreside debug              : No  (default)
* Use unix gperftools             : No  (default)
* Use GIT revision hash           : Yes (default)
* Enable vmap DisableMgr checks   : Yes (default)
* Show source tree                : No (For UNIX default)
```

--- a/src/AutoBalance.cpp
+++ b/src/AutoBalance.cpp
@@ -50,36 +50,55 @@
 
 using namespace Acore::ChatCommands;
 
-ABScriptMgr* ABScriptMgr::instance() {
+ABScriptMgr* ABScriptMgr::instance()
+{
     static ABScriptMgr instance;
     return &instance;
 }
 
-bool ABScriptMgr::OnBeforeModifyAttributes(Creature *creature, uint32 & instancePlayerCount) {
-    bool ret=true;
-    FOR_SCRIPTS_RET(ABModuleScript, itr, end, ret) // return true by default if not scripts
-    if (!itr->second->OnBeforeModifyAttributes(creature, instancePlayerCount)) {
-        ret=false; // we change ret value only when scripts return false
+bool ABScriptMgr::OnBeforeModifyAttributes(Creature *creature, uint32 & instancePlayerCount)
+{
+    auto ret = IsValidBoolScript<ABModuleScript>([&](ABModuleScript* script)
+    {
+        return !script->OnBeforeModifyAttributes(creature, instancePlayerCount);
+    });
+
+    if (ret && *ret)
+    {
+        return false;
     }
-    return ret;
+
+    return true;
 }
 
-bool ABScriptMgr::OnAfterDefaultMultiplier(Creature *creature, float & defaultMultiplier) {
-    bool ret=true;
-    FOR_SCRIPTS_RET(ABModuleScript, itr, end, ret) // return true by default if not scripts
-    if (!itr->second->OnAfterDefaultMultiplier(creature, defaultMultiplier)) {
-        ret=false; // we change ret value only when scripts return false
+bool ABScriptMgr::OnAfterDefaultMultiplier(Creature *creature, float& defaultMultiplier)
+{
+    auto ret = IsValidBoolScript<ABModuleScript>([&](ABModuleScript* script)
+    {
+        return !script->OnAfterDefaultMultiplier(creature, defaultMultiplier);
+    });
+
+    if (ret && *ret)
+    {
+        return false;
     }
-    return ret;
+
+    return true;
 }
 
-bool ABScriptMgr::OnBeforeUpdateStats(Creature* creature, uint32& scaledHealth, uint32& scaledMana, float& damageMultiplier, uint32& newBaseArmor) {
-    bool ret=true;
-    FOR_SCRIPTS_RET(ABModuleScript, itr, end, ret)
-    if (!itr->second->OnBeforeUpdateStats(creature, scaledHealth, scaledMana, damageMultiplier, newBaseArmor)) {
-        ret=false;
+bool ABScriptMgr::OnBeforeUpdateStats(Creature* creature, uint32& scaledHealth, uint32& scaledMana, float& damageMultiplier, uint32& newBaseArmor)
+{
+    auto ret = IsValidBoolScript<ABModuleScript>([&](ABModuleScript* script)
+    {
+        return !script->OnBeforeUpdateStats(creature, scaledHealth, scaledMana, damageMultiplier, newBaseArmor);
+    });
+
+    if (ret && *ret)
+    {
+        return false;
     }
-    return ret;
+
+    return true;
 }
 
 ABModuleScript::ABModuleScript(const char* name)

